internal/controller/factory/builders.go

```diff
@@ -0,0 +1,49 @@
+/*
+Copyright 2024 The etcd-operator Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package factory
+
+import (
+	"context"
+	"fmt"
+
+	appsv1 "k8s.io/api/apps/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+func reconcileSTS(ctx context.Context, rclient client.Client, crdName string, sts *appsv1.StatefulSet) error {
+	logger := log.FromContext(ctx)
+
+	currentSts := &appsv1.StatefulSet{}
+	err := rclient.Get(ctx, types.NamespacedName{Namespace: sts.Namespace, Name: sts.Name}, currentSts)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			logger.V(2).Info("creating new statefulset", "sts_name", sts.Name, "crd_object", crdName)
+			return rclient.Create(ctx, sts)
+		}
+		return fmt.Errorf("cannot get existing statefulset: %s, for crd_object: %s, err: %w", sts.Name, crdName, err)
+	}
+	sts.Annotations = labels.Merge(sts.Annotations, sts.Annotations)
+	if sts.ResourceVersion != "" {
+		sts.ResourceVersion = currentSts.ResourceVersion
```

Comment:
sts is what we build, currentSts - what we got from the cluster.

Is it just a measure to protect ourselves from accidential change of resource version outside this function?
Will it be bad if somebody changes sts manually, version is update and after that we return previous version in our code? Classic race condition.

If I got it right and it is just a measure to protect ouselves, maybe we cal make it null and do not pass it at all?

Reply:
We apply upgrade so it always should be resourceVersion from current STS

If there will be a race, we will get an error on upgrade call

Reply:
I have not developed for k8s a lot, but from my other experience resourceVersion is a value that should be changed only by k8s, not by etcd or any other custom operator. This code makes an attempt to change this value. My proposal is to avoid touching it at all. Just not to pass resource version. And if it was passed from another function => make it null in the object that will be passed to k8s api.

Anyway I will resolve this comment and in case of disagreement will find recommendations from k8s if there are any.